### ISSUE TITLE
feat: Dynamic volume thresholds via TimesFM (Phase 1)

### ISSUE THREAD

**Issue by omniscient** [omniscient/markethawk#20]
## Overview

Replace the static 4× volume multiplier in the pre-market scanner with a **per-ticker, dynamically computed volume anomaly score** using [TimesFM](https://github.com/google-research/timesfm) — Google's pretrained time-series foundation model (ICML 2024).

### Problem with the current approach

```python
avg_volume_20d = sum(volumes[-20:]) / 20
if pre_market_volume < 4 * avg_volume_20d:
    continue
```

A fixed 4× multiplier treats all tickers identically. High-volatility stocks (biotech, meme stocks) generate false positives; low-volatility stocks with genuine early breakouts get filtered out.

---

## Concept

Instead of asking *"is today's pre-market volume > 4× the 20-day mean?"*, we ask:

> **How many standard deviations outside TimesFM's predicted volume distribution is today's pre-market volume?**

This produces a continuous `volume_anomaly_score` (z-score) that is ticker-aware and adapts to each stock's historical volatility pattern.

---

## How TimesFM fits

TimesFM is a **zero-shot** foundation model — no training or fine-tuning needed. Feed it 60 days of historical daily volume and it returns a 1-day-ahead forecast with confidence intervals (p25, p50, p75, p90).

```
Input:  [vol_day_1, vol_day_2, ..., vol_day_60]  ← already in stock_aggregates
Output: { mean, std, p25, p50, p75, p90 }        ← expected volume for today
```

Anomaly score:
```
z = (actual_pre_market_volume - predicted_mean) / predicted_std
```

A score of 2.0+ means the volume is 2 standard deviations above the model's expectation — a genuine anomaly relative to *that ticker's own baseline*.

---

### Phase 1a — metadata only (no behaviour change)
Write `volume_anomaly_score` into the existing `indicators` JSONB column on `ScannerEvent`. The 4× static threshold continues to control pass/fail. This lets us validate that high scores correlate with quality signals before changing any logic.

### Phase 1b — activate dynamic filtering
Flip a `system_config` flag (`timesfm_enabled`) to switch pass/fail from the static multiplier to the z-score threshold (default: ≥ 2.0, tunable).

---

## Files to touch

| File | Change |
|------|--------|
| `backend/app/services/timeseries_forecast.py` | **New** — TimesFM wrapper (~60 lines) |
| `backend/app/services/scanner.py` | ~3-line change in `run_pre_market_scan()` |
| `docker-compose.yml` | Add `forecast-worker` service |
| `backend/alembic/versions/` | Seed 4 rows into `system_config` |

---

## New service sketch

```python
# services/timeseries_forecast.py

def get_volume_forecast(ticker: str, volumes: list[float]) -> Optional[dict]:
    """60 daily volumes -> 1-day-ahead forecast with quantile bands."""
    log_vols = np.log1p(volumes).tolist()  # log-scale stabilises volume variance
    point_forecast, quantile_forecast = model.forecast(
        inputs=[log_vols],
        freq=[0],
        quantile_levels=[0.25, 0.50, 0.75, 0.90],
    )
    # de-normalise and return { mean, std, p25, p50, p75, p90 }

def compute_anomaly_score(actual_volume: float, forecast: dict) -> float:
    """Z-score of actual vs. forecast distribution."""
    return (actual_volume - forecast['mean']) / forecast['std']
```

### Scanner change (Phase 1a)

```python
forecast = get_volume_forecast(ticker, volumes[-60:])
anomaly_score = compute_anomaly_score(pre_market_volume, forecast) if forecast else None

# Metadata only — no pass/fail change yet
indicators['volume_anomaly_score'] = anomaly_score
indicators['predicted_volume_p50'] = forecast['p50'] if forecast else None
indicators['predicted_volume_p90'] = forecast['p90'] if forecast else None
indicators['volume_threshold_method'] = 'timesfm' if forecast else 'static_4x'

# Existing threshold unchanged for Phase 1a
if pre_market_volume < 4 * avg_volume_20d:
    continue
```

---

## Infrastructure

TimesFM 2.5 (200M params) requires **32GB RAM**. Runs as a dedicated Celery worker on the `forecasting` queue with `--concurrency=1`. Model (~800MB) downloaded once from Hugging Face and persisted in a Docker volume.

```yaml
# docker-compose.yml addition
forecast-worker:
  build: ./backend  # Dockerfile.forecast adds: pip install timesfm[torch]
  command: celery -A app.celery_app worker -Q forecasting --concurrency=1
  deploy:
    resources:
      limits:
        memory: 36G
  volumes:
    - timesfm_cache:/root/.cache/huggingface
```

---

## Config knobs (system_config rows)

| Key | Default | Description |
|-----|---------|-------------|
| `timesfm_enabled` | `false` | Phase 1a: off. Flip to `true` for 1b |
| `timesfm_anomaly_threshold` | `2.0` | Min z-score to pass volume filter (Phase 1b) |
| `timesfm_min_history_bars` | `30` | Min daily bars required to run model |
| `timesfm_fallback_multiplier` | `4.0` | Static multiplier when model unavailable |

---

## What stays the same
- `stock_aggregates` schema — no changes
- `ScannerEvent` schema — no changes (score goes in existing `indicators` JSONB)
- All other scan criteria (gap %, min pre-market volume, ATR filters, etc.)
- Static 4× threshold as permanent fallback when model is unavailable

---

## References
- [TimesFM GitHub](https://github.com/google-research/timesfm)
- [TimesFM 2.5 — Hugging Face](https://huggingface.co/google/timesfm-2.5-200m-pytorch)
- [Google Research blog post](https://research.google/blog/a-decoder-only-foundation-model-for-time-series-forecasting/)

**Comment by omniscient** [comment]
## Dark Factory Run — New implementation

**PR:** #50

**Branch:** `feat/issue-20-feat--dynamic-volume-thresholds-via-time`

### Changes

# Implementation Summary: Dynamic Volume Thresholds via TimesFM (Issue #20)

## Files Created

| File | Purpose |
|------|---------|
| `backend/app/services/timeseries_forecast.py` | TimesFM wrapper with lazy-load, graceful fallback, and pure `compute_anomaly_score` |
| `backend/tests/services/test_timeseries_forecast.py` | 10 unit tests (mocked model, unavailable model, math) |
| `backend/app/alembic/versions/b5e6f7a8b9c0_seed_timesfm_system_config.py` | Seeds 4 `system_config` rows (idempotent via ON CONFLICT DO NOTHING) |
| `backend/Dockerfile.forecast` | Backend image + `pip install timesfm[torch]` for the forecast worker |

## Files Modified

| File | Change |
|------|--------|
| `backend/app/services/scanner.py` | Added SystemConfig + timeseries_forecast imports; reads 4 config keys once per scan; computes forecast + z-score per ticker; Phase 1b dynamic threshold; 4 new indicator fields |
| `backend/tests/services/test_scanner_refactor.py` | Updated `_mock_db_for_pre_market` to handle SystemConfig queries; 3 new test cases |
| `docker-compose.yml` | Added `forecast-worker` service (forecasting profile, 36GB RAM limit, timesfm_cache volume) |

## New Indicators in ScannerEvent.indicators JSONB

| Key | Description |
|-----|-------------|
| `volume_anomaly_score` | Z-score vs. TimesFM predicted distribution (None when model unavailable) |
| `predicted_volume_p50` | Median predicted volume from TimesFM (raw shares) |
| `predicted_volume_p90` | 90th-percentile predicted volume |
| `volume_threshold_method` | `"timesfm"` or `"static_4x"` — which path determined pass/fail |

## System Config Keys Seeded

| Key | Default | Effect |
|-----|---------|--------|
| `timesfm_enabled` | `false` | Phase 1a off; flip to `true` to activate Phase 1b |
| `timesfm_anomaly_threshold` | `2.0` | Min z-score to pass volume filter |
| `timesfm_min_history_bars` | `30` | Min daily bars required before calling model |
| `timesfm_fallback_multiplier` | `4.0` | Static multiplier when model path is inactive |

## Key Decisions

1. **Lazy load with permanent failure cache** — `_timesfm_available` is set to `False` after the first failed `import timesfm`. Subsequent calls skip the import entirely, adding zero latency to scans.
2. **No schema change** — score goes into existing `indicators` JSONB; no migration needed for the ScannerEvent table.
3. **Fallback preserves existing behaviour** — with `timesfm_enabled=false` (default) and TimesFM not installed, the scanner is byte-for-byte equivalent to the previous version except for the extra None/static_4x indicator fields.
4. **forecast-worker behind a profile** — `docker compose up` won't start it by default; requires `--profile forecasting`. Avoids surprises on machines without 32GB RAM.

## Test Results

- 10/10 `test_timeseries_forecast` tests pass
- 9/9 `test_scanner_refactor` tests pass (including 3 new Phase 1a/1b cases)
- 246/246 total backend tests pass (pre-existing `test_futures.py` import error unrelated to this branch)

### Preview Environment

| Service | URL |
|---------|-----|
| Frontend | http://localhost:12033 |
| Backend API | http://localhost:12080 |
| API Docs | http://localhost:12080/docs |
| PostgreSQL | `localhost:12054` |
| Redis | `localhost:12063` |

### Commands
```bash
# Iterate after review
docker compose --profile factory run --rm dark-factory "Continue issue #20"

# Tear down when done
docker compose --profile factory run --rm dark-factory "Close issue #20"
```

---
*Posted by MarketHawk Dark Factory*

**Comment by omniscient** [comment]
Fix the import error unrelated to this branch error in the tests.

**Comment by omniscient** [comment]
## Dark Factory Run — Continued iteration

**PR:** #50

**Branch:** `feat/issue-20-feat--dynamic-volume-thresholds-via-time`

### Changes

# Implementation Summary — Issue #20 Continue (fix import error)

## Feedback addressed
User asked: "Fix the import error unrelated to this branch error in the tests."

## Root cause
`tests/api/test_futures.py` imported `mock_futures_provider` from `tests.fixtures.providers`,
but that fixture was never defined there — only `mock_polygon_provider` and `mock_news_provider`
existed.

## Fix
Added `mock_futures_provider` fixture to `backend/tests/fixtures/providers.py`:
- Replaces the `ibkr` entry in `DataProviderFactory._providers` with a MagicMock
- Reports `is_available() = (True, "Ready (mock)")` and `supported_asset_classes = ["futures"]`
- Restores the original provider on teardown (same pattern as `mock_polygon_provider`)

## File modified
| File | Change |
|------|--------|
| `backend/tests/fixtures/providers.py` | Added `mock_futures_provider` fixture (+22 lines) |

## Test results
- `tests/api/test_futures.py`: 10/10 PASSED (previously 0 collected due to ImportError)
- Full suite: 256/256 PASSED

### Preview Environment

| Service | URL |
|---------|-----|
| Frontend | http://localhost:12033 |
| Backend API | http://localhost:12080 |
| API Docs | http://localhost:12080/docs |
| PostgreSQL | `localhost:12054` |
| Redis | `localhost:12063` |

### Commands
```bash
# Iterate after review
docker compose --profile factory run --rm dark-factory "Continue issue #20"

# Tear down when done
docker compose --profile factory run --rm dark-factory "Close issue #20"
```

---
*Posted by MarketHawk Dark Factory*

**Comment by omniscient** [comment]
Looks good

**Comment by omniscient** [comment]
Merging PR and tearing down preview environment.

**Comment by omniscient** [comment]
Dark factory: preview torn down, PR #50 merged. Issue complete.

**Comment by omniscient** [comment]
Merging PR and tearing down preview environment.
